lib/snippets/kubernetes-deploy

```diff
@@ -75,90 +203,375 @@ class KubernetesDeploy
     end
 
-    _, err, st = run_kubectl("config", "use-context", @context)
+    _, err, st = run_kubectl("config", "use-context", @context, namespaced: false)
     raise FatalDeploymentError, "Kubectl config is not valid: #{err}" unless st.success?
-    logger.info("Kubectl configured")
+    KubernetesDeploy.logger.info("Kubectl configured to use context #{@context}")
   end
 
   def validate_namespace
-    _, _, st = run_kubectl("get", "namespace", @namespace)
+    _, _, st = run_kubectl("get", "namespace", @namespace, namespaced: false)
     raise FatalDeploymentError, "Failed to validate namespace #{@namespace}" unless st.success?
-    logger.info("Namespace validated")
+    KubernetesDeploy.logger.info("Namespace #{@namespace} validated")
   end
 
-  def run_kubectl(*args)
+  def run_kubectl(*args, namespaced: true)
     args = args.unshift("kubectl")
-    logger.debug Shellwords.join(args)
+    if namespaced
+      raise FatalDeploymentError, "Namespace missing for namespaced command" unless @namespace
+      args.push("--namespace=#{@namespace}")
+    end
+    KubernetesDeploy.logger.debug Shellwords.join(args)
     out, err, st = Open3.capture3(*args)
-    logger.debug(out.shellescape)
-    logger.warn(err) unless st.success?
+    KubernetesDeploy.logger.debug(out.shellescape)
+    KubernetesDeploy.logger.warn(err) unless st.success?
     [out.chomp, err.chomp, st]
   end
 
-  def logger
+  def phase_heading(phase_name)
+    @current_phase += 1
+    heading = "Phase #{@current_phase}: #{phase_name}"
+    padding = (100.0 - heading.length)/2
+    KubernetesDeploy.logger.info("")
+    KubernetesDeploy.logger.info("#{'-' * padding.floor}#{heading}#{'-' * padding.ceil}")
+  end
+
+  def log_green(msg)
+    STDOUT.puts "\033[0;32m#{msg}\x1b[0m\n" # green
+  end
+
+  def self.logger
     @logger ||= begin
       l = Logger.new(STDOUT)
       l.level = ENV["DEBUG"] ? Logger::DEBUG : Logger::INFO
       l.formatter = proc do |severity, _datetime, _progname, msg|
         case severity
-        when "FATAL" then "\033[0;31m[#{severity}]\t#{msg}\x1b[0m\n"
-        when "ERROR", "WARN" then "\033[0;31m[#{severity}]\t#{msg}\x1b[0m\n"
-        when "INFO" then "\033[0;36m#{msg}\x1b[0m\n"
+        when "FATAL" then "\033[0;31m[#{severity}]\t#{msg}\x1b[0m\n" # red
+        when "ERROR", "WARN" then "\033[0;33m[#{severity}]\t#{msg}\x1b[0m\n" # yellow
+        when "INFO" then "\033[0;36m#{msg}\x1b[0m\n" # blue
         else "[#{severity}]\t#{msg}\n"
         end
       end
       l
     end
   end
+
+  class KubernetesResource
+    attr_reader :name, :namespace, :file
+    attr_writer :type
+
+    # Source: Rails
+    # https://github.com/rails/rails/blob/b292b76c2dd0f04fb090d49b90716a0e6037b41a/activesupport/lib/active_support/core_ext/class/subclasses.rb#L22-L29
+    def self.descendants
+      descendants = []
+      ObjectSpace.each_object(singleton_class) do |k|
```

Comment:
This implementation is good for had hock inspection of a class. But when you know you'll need to call it, there is a better implementation: https://github.com/rails/rails/blob/9f84e60ac9d7bf07d6ae1bc94f3941f5b8f1a228/activesupport/lib/active_support/descendants_tracker.rb